### ISSUE TITLE
[Feature Request] Granular warning messages

### ISSUE THREAD

**Issue by amogorkon** [beartype/beartype#73]
I would like to apply beartype as an aspect-decorator via justuse
`use(thing) @ (isfunction, "", beartype)`
but with only
```
g:\Python398\lib\site-packages\beartype\_util\hint\pep\utilpeptest.py:373: BeartypeDecorHintPep585DeprecationWarning: PEP 484 type hint typing.Callable[..., 
typing.Any] deprecated by PEP 585 scheduled for removal in the first Python version released after October 5th, 2025. To resolve this, either drop Python < 3.9 support and globally replace this hint by the equivalent PEP 585 type hint (e.g., "typing.List[int]" by "list[int]") or see this discussion topic for saner and more portable solutions:
    https://github.com/beartype/beartype#pep-484-deprecations
  warn(
```
as warning, there's not much to go on..

**Comment by leycec** [comment]
**You're absolutely right.** We'd really love for `justuse` to pound everything into the ground with `beartype`. We'll make this happen faster than you can say, "Bug-free is the way to be."

Interestingly, this exact issue is something my younger self continually complained about in `FIXME:` comments littered throughout the codebase. My older self decided to just binge [*The Stormlight Archive*](https://www.goodreads.com/book/show/7235533-the-way-of-kings) instead. Here we are.

## Just Fix This, Please.

Interestingly, [this is also a common world-straddling complaint throughout the Python world](http://pymotw.com/2/warnings/#generating-warnings). Whereas exceptions come bundled with traceback context out-of-the-box, warnings sadly don't. Until we resolve this on our end, would you mind trying one of the following end-user workarounds? You can either:

* **Enable the `error` warnings filter programmatically.** The standard `warnings` module enables you to trivially coerce *all* warnings of a certain category into corresponding exceptions. This has the disadvantage of raising and burning everything to the ground, but the advantage of emitting exception tracebacks while doing so. For granularity, you probably only want to coerce warnings of either the high-level `beartype.roar.BeartypeWarning` *or* low-level `beartype.roar.BeartypeDecorHintPep585DeprecationWarning` categories to exceptions. Lastly, you should ideally do this *before* decorating any callables with the `@beartype` decorator:

``` python
from beartype import beartype
from beartype.roar import BeartypeDecorHintPep585DeprecationWarning
from warnings import simplefilter

# Coerce all PEP 585 deprecation warnings into fatal exceptions.
simplefilter('error', BeartypeDecorHintPep585DeprecationWarning)

# Under Python ≥ 3.9, this should now raise a fatal exception.
from typing import Callable, List
@beartype
def isfunctions(funcs: List[Callable]) -> bool:
    return all(callable(func) for func in funcs)
```

* Alternately, **enable the `error` warnings filter from the command line.** You can also archeive The from the CLI via the standard `-W` option accepted by *all* `python` binaries. I kinda doubt this approach supports dynamic importation of third-party warning categories. Thankfully, ours all subclass `UserWarning`. This should thus do you as well (albeit somewhat more ambiguously): e.g.,

``` bash
$ python -W "error::UserWarning::0" muh_script.py
```

** Lastly, **enable the `error` warnings filter declaratively from your test suite.** `pytest`, for example, supports this out-of-the-box with the `filterwarnings` setting in the top-level `pytest.ini` configuration file. Here at `beartype`, [we unconditionally coerce *all* warnings into errors to guarantee both test failures and exception tracebacks on all warnings](https://github.com/beartype/beartype/blob/3df840dfd11f09dc9b6f04cb6d29703909e2cb0a/pytest.ini#L24) with a `filterwarnings` setting resembling:

```
filterwarnings =
    error
```

Personally, we favour the test-driven approach. We repeatedly faceplanted into this issue *again* and *again* with various third-party frameworks (especially Big Scary Ones (BSOs) like NumPy and SciPy). So, globally nipping this in the bud seemed like the sanest approach.

## Tschüß, Ami!

Thanks yet again for all the incisive suggestions, Anselm.

Completely unrelatedly, I loved and adored my brief stay in Germany. What a delightful nation you've all made there! I dabbled in AI research at the [ Universität Bremen's Technologie-zentrum informatik (TZI)](https://www.uni-bremen.de/tzi). Ahh – to be foolishly young and guzzling buckets of German lager in the summertime tiergartens of Bremen again. :sun_with_face: :beers:

**Comment by leycec** [comment]
Oh, and I hope you don't mind my lightly massaging the issue title. As GitHub stars and issues alike explode in volume, I'm in triage mode sifting the "This would be *really* great to have!" from the "The stars are burning pits of fission and so is my code because of you."

That said, I *really* want this to happen. NumPy-style contextless warnings are nobody's friend. :disappointed:

**Comment by amogorkon** [comment]
So,  we've investigated where those beartype warnings were coming from.

Let me explain a little about the context. 
When you use() something, a ProxyModule is returned, which adds extra functionality to that module like `__matmul__` which lets you decorate *everything* in a module that passes a check (inspect.isfunction for example) and whose name matches the given regex ("" matches everything).
We use this also in our code to beartype everything and help log/debug stuff.

In this case, I guess our smartness turned a little against us. Aspectizing our own code worked quite well up until recently when those beartype warnings popped up and had us ~hex~ "refine"  built-in types, solving the problem for our own code. However, since aspectizing *everything* also includes imports, at one point we had over 800 beartype warnings coming from dependencies, so we started filtering based on point of origin and we could get rid of all but 2 warnings.

Now the ironic conclusion: Turns out we were applying beartype to beartype - which was complaining about itself, making us chase our own tails ^_^"

PS: AI research in Bremen sounds absolutely lovely! Although, I fear corona is quite the spoilsport these days again - we're on "lockdown for non-immunized people" mode now, about to ruin xmas 3 times in a row now.

**Comment by leycec** [comment]
**Wow.** Because `justuse` is endlessly fascinating, I have an endless reel of tiresome questions that will exhaust you even as they fire you up. Let's begin.

First, I *love* that [you have a public submodule named `pimp` that gently massages `pip`](https://github.com/amogorkon/justuse/blob/689fe0ff1f8efd1a1eb7c5c0eff9202215dfb453/src/use/pimp.py). Also, this docstring commentary on life in the packaging trenches:

```
    Results of this function are cached. They are expensive to
    compute, thanks to some heavyweight usual players
    (*ahem* pip, package_resources, packaging.tags *cough*)
    whose modules are notoriously resource-hungry.
```

*heh*

Second, pinch me I must be dreaming, but did you *really* just imply that `justuse` is a working aspect-oriented meta-decorator that can be used to transitively apply any arbitrary decorator to all callables declared under any arbitrary package? Specifically, this:

> When you `use()` something, a `ProxyModule` is returned, which adds extra functionality to that module... which lets you decorate everything in a module that passes a check (`inspect.isfunction` for example) and whose name matches the given regex (`""` matches everything). We use this also in our code to `beartype` everything and help log/debug stuff.

![conspiracy keanu](http://www.quickmeme.com/img/2d/2df73b4f8f523d4b2b95bc76860c9d320dbe68ca8d8014ea3725fe4fad569644.jpg)

If true, that's an impressive *and* (more importantly) immediate solution to the [problem of globally applying `@beartype` across an entire package that everyone wants us to solve with import hook automation](https://github.com/beartype/beartype/issues/43). I regretfully lack the time to solve that at the moment, because laziness is a virtue. That's where you enter stage right. If you've already solved that for us, I can quietly steer everyone towards you and say:

> Just use `justuse`. <sup>micdrop</sup>

But... how *exactly* should everyone use `justuse`? I tried grepping the main branch of the `justuse` codebase for `beartype` and came up mostly empty-pawed. I also tried crawling through your excellent documentation, but then I got tired and stopped. The sleep deprivation is real.

When you get a spare moment, would you mind replying with a working minimal-length example (MLE) that globally applies `@beartype` to an arbitrary package? I'll then giddily document this all over our FAQ. Your GitHub karma will shoot to the moon and beyond. Soon, rabid `justuse` fanpersons will be doxxing a random stranger in Bulgaria who looks suspiciously like your GitHub avatar. I say: *just let them.*

We'll still do import hook automation *eventually*, of course. Because it's cool and scary and therefore attractive to risk-prone code slingers like us. Can't help but meddle with the dark side, can we?

> Turns out we were applying beartype to beartype - which was complaining about itself, making us chase our own tails ^_^

*heh*

**...wait.** That's actually superbad. I *never* thought of going recursive – but now that you've unsealed Pandora's Snuff Box, I can't help but feel that we're hypocritical. We warn everyone else not to use PEP 484 type hints under Python ≥ 3.9... *and then we do exactly that.*

Honestly, I was kinda hoping nobody would notice that. Welp. This is awkward.

**Comment by amogorkon** [comment]
> Second, pinch me I must be dreaming, but did you really just imply that justuse is a working aspect-oriented meta-decorator that can be used to transitively apply any arbitrary decorator to all callables declared under any arbitrary package? 
Yes and no. It works well enough for flat modules, but I dropped the ball on deep recursion and transitivity a while back when nobody seemed to care much about aspectizing at all and things got weird on the packaging front. Would you mind [opening an issue](https://github.com/amogorkon/justuse/issues?q=is%3Aopen+is%3Aissue+milestone%3A%22Aspect-Oriented+Decoration%22) for the specific cases you're thinking of? I'd love to give it the TLC it deserves :-)

> When you get a spare moment, would you mind replying with a working minimal-length example (MLE) that globally applies @beartype to an arbitrary package?
Actually the last few days I've been working on a [Showcase](https://github.com/amogorkon/justuse/blob/unstable/docs/Showcase.ipynb), just haven't gotten to the aspectizing part just yet! Maybe it'll be done when you're awake again ;-)

**Comment by leycec** [comment]
**Yes!** You ask. [I comply.](https://github.com/amogorkon/justuse/issues/141#issuecomment-979738894) That's just how we roll here in the bear cave.

**Comment by amogorkon** [comment]
If you care to check out the [Showcase](https://github.com/amogorkon/justuse/blob/unstable/docs/Showcase.ipynb), you will find   I've added a chapter about aspectizing towards the end. I also asked my companions to help with aspectizing, so I'm optimistic we can come up with something more usable soon(tm).

**Comment by leycec** [comment]
Firstly, **that's hot.** I never knew GitHub-rendered Jupyter notebook cells could look so svelte. :open_mouth: 

I'm now considering ~~stealing~~ friendly-borrowing-with-attribution your documentation pattern for [our own modest but less well-rendered showcase](https://github.com/beartype/beartype#validator-showcase).

## C-C-Can It Be!?

Secondly:

```
import use
from beartype import beartype
from use import isfunction

np = use('numpy')
np @ (isfunction, '', beartype)
```

*zomg.* You just runtime type-checked the entirety of NumPy without its permission. When you wrote "adds extra functionality to that module like `__matmul__`" above, I boldly assumed your last two words were a copypasta error from elsewhere. You couldn't possibly be overloading the `@` operator for nefarious aspectization purposes... *could you?*

You are Anselm. What couldn't you do?

Does that terrifying yet aesthetically pleasing magic work under the most recent stable release of `justuse` (i.e., 0.6.5) or is this still on your bleeding edge? If the former, then it's time to roll up the grease-stained pant legs and get to work documenting this in our FAQ. *Dis gunna be fun.*

## Slay All the Dragons

Thirdly, this actually leads us to a topic that is completely tangential to our normals: [**PEP 563**](https://www.python.org/dev/peps/pep-0563).

I've long suspected that PEP 563 dramatically reduces the decoration-time efficiency of runtime type-checkers. Issue #71 all but confirmed that suspicion. With `justuse`, I can now quantify *exactly* how badly PEP 563 impacts decoration-time performance by profiling the wall-clock costs of:

* Aspectizing NumPy with `@beartype` **+** PEP 563 globally disabled (i.e., the default).
* Aspectizing NumPy with `@beartype` **+** PEP 563 globally enabled (e.g., by manually forcing this deep from within the festering bowels of `@beartype`).

NumPy is the perfect type-hinting torture test, because they take type hints to hitherto unimagined heights of insanity. With concrete timings in hand, I'll *finally* have something tangible to approach the Python Steering Council with. Oh, boy! Python mailing list: prepare for incoming @leycec.

**Comment by amogorkon** [comment]
1) I'm glad you like it! No need to attribute, I only repurposed already existing functionality and would be honoured if you feel inspired by it ^_^
2) WYSIWYG - I made sure to only include stuff in the showcase that worked out of the box. However, there are currently limitations as outlined in the justuse issue.
3) I have no idea what you're talking about, but it sounds like fun ;o) Just make sure that the overhead by the attribute lookup in the module proxy doesn't skew your timings.

**Comment by leycec** [comment]
> I have no idea what you're talking about, but it sounds like fun ;o)

Fun benefits from generous air quotes in this context. :face_exhaling: 

Everyone using runtime type-checkers should probably know about [PEP 563](https://www.python.org/dev/peps/pep-0484/#union-types), so... Here goes the hellshow. This is PEP 563:

```python
from __future__ import annotations
```

**That's it.** Insert that at the *very* top of any Python module as the first statement of that module and you've now conditionally enabled PEP 563 for that module. Congratulations. You have successfully destroyed runtime performance. Why? Because:

```python
# Enabling PEP 563 stringifies all type hints from actual usable objects like this...
def muh_func(muh_arg: list[str]) -> tuple[int, bool]: pass

# ...into unusable string placeholders like this.
def muh_func(muh_arg: 'list[str]') -> 'tuple[int, bool]': pass
```

Nobody can directly use a stringified type hint like `'list[str]'` or `'tuple[int, bool]'` at runtime. That means everyone now needs to dynamically reparse those stringified type hints back into their original objects like `list[str]` or `tuple[int, bool]` at runtime when PEP 563 is enabled. That's computationally expensive and algorithmically non-trivial, so everyone except `@beartype` just gave up and didn't bother trying. *Fair enough.*

PEP 563 is [the thing that almost destroyed `pydantic`, the FastAPI web framework, and everything based on either of those two things](https://github.com/samuelcolvin/pydantic/issues/2678). PEP 563 was [the thing that Python 3.10 *almost* unconditionally forced on everyone but which Python's new BDFL wisely reverted at the last second to avoid breaking the entire world](https://mail.python.org/archives/list/python-dev@python.org/thread/CLVXXPQ2T2LQ5MP2Y53VVQFCXYWQJHKZ). Pretty big thing; lotsa shouting matches; much "fun" for everyone involved in the mud-flinging scrum.

We'd all like to bury PEP 563 once and for all. If we can extract useful metrics out of `beartype` **+** `justuse` **+** NumPy, we can run to the Python Steering Council with an evidence-based argument for doing just that. In short, praise be to `justuse`, for it shall deliver us from evil. :pray:

**Comment by amogorkon** [comment]
Ohhh, THAT PEP.. yeah, I heard the screaming and shouting from afar. By that time it sounded like a good idea, having run into problems with forward references etc myself, but yeah.. having pudantic

**Comment by amogorkon** [comment]
Crap. Shouldn't comment on github issues on my mobile.

**Comment by leycec** [comment]
**ZOMG.** Back on topic, I *never* knew that [the standard `warnings.warn()` function accepted an optional `stacklevel` keyword parameter intended exactly for situations like this](https://docs.python.org/3/library/warnings.html#warnings.warn). I now feel bad, because I made @beartype emit unhelpful warnings that required non-trivial hacks from other devbros to work around. Take all my crying cat faces, GitHub! :joy_cat:  :crying_cat_face:

Of course, it's unclear exactly what `stacklevel` @beartype should be passing to `warnings.warn()`. The nesting level is *pretty* deep there. Hard-coding a fragile magic number guaranteed to silently fail on the next deep refactoring invites bad juju – and I've already incurred enough for five lifetimes. That ain't easy to work off, folks.

Let's cogitate on this while I continue doing nothing.

**Comment by leycec** [comment]
**So.** The final journey of feature request #73 begins. :city_sunset:

**Comment by leycec** [comment]
**Resolved by 438fe70.** At long last, the promised land has arrived. After two years of sullen silence and much video gaming, @beartype 0.18.0 now offers detailed context for *all* warnings – including the much-feared PEP 585 deprecation warnings:

```python
>>> from beartype import beartype
>>> from typing import List
>>> @beartype
... def oh(god: List[str]) -> str:
... return god[0]
>>> oh(['We are born of the blood', 'made men by the blood', 'undone by the blood.'])
BeartypeDecorHintPep585DeprecationWarning: Function __main__.oh() parameter "god"
PEP 484 type hint typing.List[str] deprecated by PEP 585. This hint is scheduled for
removal in the first Python version released after October 5th, 2025. To resolve this,
import this hint from "beartype.typing" rather than "typing".
```

The bear snickers even as it flexes its biceps. :muscle:  :bear:
